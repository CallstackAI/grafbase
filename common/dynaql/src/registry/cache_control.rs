/// Cache control values
///
/// # Examples
///
/// ```rust, ignore
/// use dynaql::*;
///
/// struct Query;
///
/// #[Object(cache_control(max_age = 60))]
/// impl Query {
///     #[graphql(cache_control(max_age = 30))]
///     async fn value1(&self) -> i32 {
///         0
///     }
///
///     #[graphql(cache_control(private))]
///     async fn value2(&self) -> i32 {
///         0
///     }
/// }
///
/// # tokio::runtime::Runtime::new().unwrap().block_on(async {
/// let schema = Schema::new(Query, EmptyMutation, EmptySubscription);
/// assert_eq!(schema.execute("{ value1 }").await.into_result().unwrap().cache_control, CacheControl { public: true, max_age: 30 });
/// assert_eq!(schema.execute("{ value2 }").await.into_result().unwrap().cache_control, CacheControl { public: false, max_age: 60 });
/// assert_eq!(schema.execute("{ value1 value2 }").await.into_result().unwrap().cache_control, CacheControl { public: false, max_age: 30 });
/// # });
/// ```
#[derive(Clone, Copy, PartialEq, Eq, Debug, serde::Deserialize, serde::Serialize, Hash)]
pub struct CacheControl {
    /// Scope is public, default is true.
    pub public: bool,

    /// Cache max age, default is 0.
    pub max_age: usize,

    /// Cache stale_while_revalidate, default is 0.
    #[serde(default)]
    pub stale_while_revalidate: usize,
}

impl Default for CacheControl {
    fn default() -> Self {
        Self {
            public: true,
            max_age: 0,
            stale_while_revalidate: 0,
        }
    }
}

impl CacheControl {
    #[must_use]
    pub(crate) fn merge(self, other: &CacheControl) -> CacheControl {
        CacheControl {
            public: self.public && other.public,
            max_age: if self.max_age == 0 {
                other.max_age
            } else if other.max_age == 0 {
                self.max_age
            } else {
                self.max_age.min(other.max_age)
            },
            stale_while_revalidate: if self.stale_while_revalidate == 0 {
                other.stale_while_revalidate
            } else if other.stale_while_revalidate == 0 {
                self.stale_while_revalidate
            } else {
                self.stale_while_revalidate
                    .min(other.stale_while_revalidate)
            },
        }
    }
}
