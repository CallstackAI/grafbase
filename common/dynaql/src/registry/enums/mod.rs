mod order_by_direction;
pub use order_by_direction::OrderByDirection;

use unindent::unindent;

pub struct DynaqlEnums;

impl DynaqlEnums {
    pub fn sdl() -> String {
        OrderByDirection::sdl()
    }
}

pub trait DynaqlEnum {
    fn ty() -> &'static str;
    fn values() -> Vec<String>;
    fn sdl() -> String {
        unindent(&format!(
            r#"
            enum {ty} {{
                {values}
            }}
            "#,
            ty = Self::ty(),
            values = Self::values().join("\n    ")
        ))
    }
}
