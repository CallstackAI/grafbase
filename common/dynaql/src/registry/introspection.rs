//! Allow converting a [`cynic_introspection::Schema`] into a new [`Registry`].
//!
//! The conversion takes all the known information from the introspection and adds it to the
//! registry.
//!
//! # Panics
//!
//! The conversion is infallible (using [`From`] instead of [`TryFrom`]), but is allowed to panic,
//! if a future update to [`cynic_introspection`] introduces new enum variants to any of the
//! non-exhaustive enums in the crate. This _should_ not happen, unless the actual [GraphQL
//! specification] changes.
//!
//! [GraphQL specification]: https://spec.graphql.org

use super::{
    Deprecation, MetaDirective, MetaEnumValue, MetaField, MetaInputValue, MetaType, Registry,
    __DirectiveLocation,
};

impl From<cynic_introspection::Schema> for Registry {
    fn from(schema: cynic_introspection::Schema) -> Self {
        let mut registry = Self::new();

        // root types
        registry.query_type = schema.query_type;
        registry.mutation_type = schema.mutation_type;
        registry.subscription_type = schema.subscription_type;

        // directives
        registry.directives = schema
            .directives
            .into_iter()
            .map(|d| (d.name.clone(), d.into()))
            .collect();

        // types
        registry.types = schema
            .types
            .into_iter()
            .map(|ty| {
                use cynic_introspection::Type::*;

                let key = match &ty {
                    Object(v) => v.name.clone(),
                    InputObject(v) => v.name.clone(),
                    Enum(v) => v.name.clone(),
                    Interface(v) => v.name.clone(),
                    Union(v) => v.name.clone(),
                    Scalar(v) => v.name.clone(),
                    _ => unimplemented!("unknown graphql type"),
                };

                (key, ty.into())
            })
            .collect();

        registry
    }
}

impl From<cynic_introspection::Directive> for MetaDirective {
    fn from(directive: cynic_introspection::Directive) -> Self {
        Self {
            name: directive.name,
            description: directive.description,
            locations: directive.locations.into_iter().map(Into::into).collect(),
            args: directive
                .args
                .into_iter()
                .map(|v| (v.name.clone(), v.into()))
                .collect(),
            is_repeatable: false,
            visible: None,
        }
    }
}

impl From<cynic_introspection::Field> for MetaField {
    fn from(field: cynic_introspection::Field) -> Self {
        Self {
            name: field.name,
            description: field.description,
            args: field
                .args
                .into_iter()
                .map(|v| (v.name.clone(), v.into()))
                .collect(),
            ty: field.ty.to_string(),
            deprecation: field.deprecated.into(),
            ..Default::default()
        }
    }
}

impl From<cynic_introspection::InputValue> for MetaInputValue {
    fn from(input: cynic_introspection::InputValue) -> Self {
        Self {
            name: input.name,
            description: input.description,
            ty: input.ty.name,
            default_value: input.default_value.map(Into::into),
            visible: None,
            validators: None,
            is_secret: false,
            rename: None,
        }
    }
}

impl From<cynic_introspection::DirectiveLocation> for __DirectiveLocation {
    fn from(location: cynic_introspection::DirectiveLocation) -> Self {
        use __DirectiveLocation::*;
        use cynic_introspection::DirectiveLocation::*;

        match location {
            Query => QUERY,
            Mutation => MUTATION,
            Subscription => SUBSCRIPTION,
            Field => FIELD,
            FragmentDefinition => FRAGMENT_DEFINITION,
            FragmentSpread => FRAGMENT_SPREAD,
            InlineFragment => INLINE_FRAGMENT,
            VariableDefinition => VARIABLE_DEFINITION,
            Schema => SCHEMA,
            Scalar => SCALAR,
            Object => OBJECT,
            FieldDefinition => FIELD_DEFINITION,
            ArgumentDefinition => ARGUMENT_DEFINITION,
            Interface => INTERFACE,
            Union => UNION,
            Enum => ENUM,
            EnumValue => ENUM_VALUE,
            InputObject => INPUT_OBJECT,
            InputFieldDefinition => INPUT_FIELD_DEFINITION,
        }
    }
}

impl From<cynic_introspection::Type> for MetaType {
    fn from(ty: cynic_introspection::Type) -> Self {
        use cynic_introspection::Type::*;

        match ty {
            Object(v) => v.into(),
            InputObject(v) => v.into(),
            Enum(v) => v.into(),
            Interface(v) => v.into(),
            Union(v) => v.into(),
            Scalar(v) => v.into(),
            _ => unimplemented!("unknown graphql type"),
        }
    }
}

impl From<cynic_introspection::ObjectType> for MetaType {
    fn from(object: cynic_introspection::ObjectType) -> Self {
        Self::Object {
            name: object.name,
            description: object.description,
            fields: object
                .fields
                .into_iter()
                .map(|v| (v.name.clone(), v.into()))
                .collect(),
            cache_control: Default::default(),
            extends: false,
            keys: None,
            visible: None,
            is_subscription: false,
            is_node: false,
            rust_typename: String::new(),
            constraints: vec![],
        }
    }
}

impl From<cynic_introspection::Deprecated> for Deprecation {
    fn from(deprecated: cynic_introspection::Deprecated) -> Self {
        use cynic_introspection::Deprecated::*;
        use Deprecation::*;

        match deprecated {
            No => NoDeprecated,
            Yes(reason) => Deprecated { reason },
            _ => unimplemented!("unknown graphql deprecation type"),
        }
    }
}

impl From<cynic_introspection::InputObjectType> for MetaType {
    fn from(input: cynic_introspection::InputObjectType) -> Self {
        Self::InputObject {
            name: input.name,
            description: input.description,
            input_fields: input
                .fields
                .into_iter()
                .map(|v| (v.name.clone(), v.into()))
                .collect(),
            visible: None,
            rust_typename: String::new(),
            oneof: false,
        }
    }
}

impl From<cynic_introspection::EnumType> for MetaType {
    fn from(enum_type: cynic_introspection::EnumType) -> Self {
        Self::Enum {
            name: enum_type.name,
            description: enum_type.description,
            enum_values: enum_type
                .values
                .into_iter()
                .map(|v| (v.name.clone(), v.into()))
                .collect(),
            visible: None,
            rust_typename: String::new(),
        }
    }
}

impl From<cynic_introspection::EnumValue> for MetaEnumValue {
    fn from(enum_value: cynic_introspection::EnumValue) -> Self {
        Self {
            name: enum_value.name,
            description: enum_value.description,
            deprecation: enum_value.deprecated.into(),
            visible: None,
            value: None,
        }
    }
}

impl From<cynic_introspection::InterfaceType> for MetaType {
    fn from(interface: cynic_introspection::InterfaceType) -> Self {
        Self::Interface {
            name: interface.name,
            description: interface.description,
            fields: interface
                .fields
                .into_iter()
                .map(|v| (v.name.clone(), v.into()))
                .collect(),
            possible_types: interface.possible_types.into_iter().collect(),
            extends: false,
            keys: None,
            visible: None,
            rust_typename: String::new(),
        }
    }
}

impl From<cynic_introspection::UnionType> for MetaType {
    fn from(union: cynic_introspection::UnionType) -> Self {
        Self::Union {
            name: union.name,
            description: union.description,
            possible_types: union.possible_types.into_iter().collect(),
            visible: None,
            rust_typename: String::new(),
            discriminators: None,
        }
    }
}

impl From<cynic_introspection::ScalarType> for MetaType {
    fn from(scalar: cynic_introspection::ScalarType) -> Self {
        Self::Scalar {
            name: scalar.name,
            description: scalar.description,
            is_valid: None,
            visible: None,
            specified_by_url: None,
        }
    }
}

#[cfg(test)]
mod tests {
    use cynic_introspection::query::IntrospectionQuery;

    use super::*;

    #[test]
    fn conversion() {
        let data = include_str!("../../tests/swapi_introspection.json");
        let schema = serde_json::from_str::<IntrospectionQuery>(data)
            .unwrap()
            .into_schema()
            .unwrap();

        insta::assert_debug_snapshot!(Registry::from(schema))
    }
}
