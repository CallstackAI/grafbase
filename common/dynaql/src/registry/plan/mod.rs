use query_planning::logical_plan::apply_fct::ApplyFunction;
use serde::{Deserialize, Serialize};

use super::SchemaID;

#[non_exhaustive]
#[derive(Clone, Debug, Deserialize, Serialize)]
pub enum SchemaPlan {
    /// Create a projection by selecting only a subset of the data.
    Projection(PlanProjection),
    /// Get the Entities related to this node
    Related(PlanRelated),
    /// Apply a function on a column
    Apply(Apply),
    First(First),
    Last(Last),
    PaginationPage(PaginationPage),
    Resolver(Resolver),
}

impl SchemaPlan {
    pub fn is_from_maindb(&self) -> bool {
        match self {
            SchemaPlan::Projection(_) => true, // We don't know in fact.
            SchemaPlan::Related(_) => true,
            SchemaPlan::Apply(ref input) => input.plan.as_ref().is_from_maindb(),
            SchemaPlan::First(ref input) => input
                .plan
                .as_ref()
                .map(|x| x.is_from_maindb())
                .unwrap_or(true),
            SchemaPlan::Last(ref input) => input
                .plan
                .as_ref()
                .map(|x| x.is_from_maindb())
                .unwrap_or(true),
            SchemaPlan::PaginationPage(_) => true,
            SchemaPlan::Resolver(_) => false,
        }
    }
}

impl SchemaPlan {
    pub fn projection(fields: Vec<String>) -> Self {
        Self::Projection(PlanProjection { fields })
    }

    pub fn resolver(resolver_name: String) -> Self {
        Self::Resolver(Resolver { resolver_name })
    }

    pub fn related(
        from: Option<SchemaID>,
        to: SchemaID,
        relation_name: Option<String>,
        ty: String,
    ) -> Self {
        Self::Related(PlanRelated {
            from,
            to,
            relation_name,
            ty,
        })
    }

    pub fn apply_cursor_encode(self, fields: Vec<String>) -> Self {
        Self::Apply(Apply {
            plan: Box::new(self),
            fun_fields: fields
                .into_iter()
                .map(|field| (field, ApplyFunction::CursorEncode))
                .collect(),
        })
    }

    pub fn first(previous: Option<Self>) -> Self {
        Self::First(First {
            plan: previous.map(Box::new),
        })
    }

    pub fn last(previous: Option<Self>) -> Self {
        Self::Last(Last {
            plan: previous.map(Box::new),
        })
    }

    pub fn pagination_page(page: PaginationPage) -> Self {
        Self::PaginationPage(page)
    }
}

/// Describe the fields projected
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct PlanProjection {
    pub(crate) fields: Vec<String>,
}

/// Describe the relation
/// TODO: When handling Union for GraphQL: We need to sort an Union of multiple Schema.
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct PlanRelated {
    pub(crate) from: Option<SchemaID>,
    pub(crate) to: SchemaID,
    pub(crate) relation_name: Option<String>,
    /// Type name for the output Schema.
    pub(crate) ty: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Apply {
    pub plan: Box<SchemaPlan>,
    pub(crate) fun_fields: Vec<(String, ApplyFunction)>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct First {
    pub plan: Option<Box<SchemaPlan>>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Last {
    pub plan: Option<Box<SchemaPlan>>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub enum PaginationPage {
    Next,
    Previous,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Resolver {
    pub resolver_name: String,
}
