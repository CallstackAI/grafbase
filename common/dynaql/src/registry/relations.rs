use dynaql_parser::types::{BaseType, Type};

#[derive(Clone, PartialEq, Eq, Debug, serde::Deserialize, serde::Serialize, Hash)]
pub enum MetaRelationKind {
    ManyToMany,
    ManyToOne,
    OneToMany,
    OneToOne,
}

// TODO: Add a link to the documentation related to this.
#[derive(Debug, thiserror::Error)]
pub enum RelationCombinationError {
    #[error(
        "You have an issue while modelizeing your relations. Try using the `@relation` directive."
    )]
    UndefinedError,
    #[error("You have multiple relations starting from `{from}`, the relation engine can't define them without a little help. Try using the `@relation` directive.")]
    MultipleRelationsError { from: String },
}

impl MetaRelationKind {
    #[allow(dead_code)]
    const fn inverse(&self) -> Self {
        match &self {
            Self::ManyToMany => Self::ManyToMany,
            Self::OneToMany => Self::ManyToOne,
            Self::ManyToOne => Self::OneToMany,
            Self::OneToOne => Self::OneToOne,
        }
    }

    pub fn new(from: &Type, to: &Type) -> Self {
        match (&from.base, &to.base) {
            (BaseType::Named(_), BaseType::Named(_)) => Self::OneToOne,
            (BaseType::Named(_), BaseType::List(_)) => Self::OneToMany,
            (BaseType::List(_), BaseType::Named(_)) => Self::ManyToOne,
            (BaseType::List(_), BaseType::List(_)) => Self::ManyToMany,
        }
    }
}

/// Define relations across types.
///
/// # Mono-directional relation
///
/// In case of mono-directional relation, for instance:
///
/// Node A -> Node B
///
/// The Relation associated to that will always be a ManyToOne, `A* -> 1 B`.
/// The inner `Type` will represent the next edge of the relation.
///
/// NB: In case of Mono-directional relation, adding the other edge of the relation
/// in the GraphQL schema won't add the relation from the other direction, why?
/// Because if it's a mono-directional relation, we do not store the other edge
/// of the relation.
///
/// # Example
///
/// ## OneToOne
///
/// ### Author OneToOne Post ->
///
/// ```graphql
/// type Author @modelized {
///   published: Post
/// }
///
/// type Post @modelized {
///   ... (Not connected)
/// }
/// ```
///
/// ### Author OneToOne Post <->
///
/// ```graphql
/// type Author @modelized {
///   published: Post
/// }
///
/// type Post @modelized {
///   author: Author
/// }
/// ```
///
/// ## OneToMany
///
/// ### Author OneToMany Post ->
///
/// ```graphql
/// type Author @modelized {
///   published: [Post]
/// }
///
/// type Post @modelized {
///   ... (Not connected)
/// }
/// ```
///
/// ### Author OneToMany Post <->
///
/// ```graphql
/// type Author @modelized {
///   published: [Post]
/// }
///
/// type Post @modelized {
///   author: Author
/// }
/// ```
///
/// ## ManyToOne
///
/// ### Author ManyToOne Post ->
///
/// ```graphql
/// type Author @modelized {
///   published: Post
/// }
///
/// type Post @modelized {
///   ... (Not connected)
/// }
/// ```
///
/// ### Author ManyToOne Post <->
///
/// ```graphql
/// type Author @modelized {
///   published: Post
/// }
///
/// type Post @modelized {
///   authors: [Author]
/// }
/// ```
///
/// ## ManyToMany
///
/// ### Author ManyToMany Post ->
///
/// ```graphql
/// type Author @modelized {
///   published: [Post]
/// }
///
/// type Post @modelized {
///   ... (Not connected)
/// }
/// ```
///
/// ### Author ManyToMany Post <->
///
/// ```graphql
/// type Author @modelized {
///   published: [Post]
/// }
///
/// type Post @modelized {
///   authors: [Author]
/// }
/// ```
#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, Hash, PartialEq, Eq)]
pub struct MetaRelation {
    pub name: String,
    pub kind: MetaRelationKind,
    /// The direction is:
    /// 0 -> 1
    pub relation: (String, String),
    pub birectional: bool,
}

impl MetaRelation {
    fn generate_relation_name(a: &str, b: &str) -> String {
        let mut a = vec![a, b];
        a.sort_unstable();
        a.join("Relate")
    }

    pub fn new(name: Option<String>, from: &Type, to: &Type) -> Self {
        let base_from = from.base.to_base_type_str();
        let base_to = to.base.to_base_type_str();
        Self {
            name: name
                .unwrap_or_else(|| MetaRelation::generate_relation_name(&base_from, &base_to)),
            relation: (base_from.to_string(), base_to.to_string()),
            birectional: false,
            kind: MetaRelationKind::new(&from, &to),
        }
    }

    /// Compose the relation with another relation based on the realtion's name
    pub fn with(&mut self, relation: MetaRelation) -> Result<(), RelationCombinationError> {
        // The relation start with a OneToX, it's obligated.
        if self.name != relation.name {
            return Err(RelationCombinationError::UndefinedError);
        }

        if self.relation.0 != relation.relation.1 || self.relation.1 != relation.relation.0 {
            return Err(RelationCombinationError::MultipleRelationsError {
                from: self.relation.0.clone(),
            });
        }

        // TODO: Need to redefine this
        self.kind = match (&self.kind, relation.kind) {
            (MetaRelationKind::OneToOne, MetaRelationKind::OneToOne) => MetaRelationKind::OneToOne,
            (MetaRelationKind::OneToOne, MetaRelationKind::OneToMany) => {
                MetaRelationKind::ManyToOne
            }
            (MetaRelationKind::OneToMany, MetaRelationKind::OneToOne) => {
                MetaRelationKind::OneToMany
            }
            (MetaRelationKind::OneToMany, MetaRelationKind::OneToMany) => {
                MetaRelationKind::ManyToMany
            }
            _ => {
                return Err(RelationCombinationError::UndefinedError);
            }
        };

        self.birectional = true;
        Ok(())
    }
}
