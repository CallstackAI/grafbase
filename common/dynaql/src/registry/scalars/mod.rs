use std::str::FromStr;

use chrono::{DateTime, Utc};
use const_format::concatcp;
use dynaql_value::ConstValue;
use fast_chemail::parse_email;
use url::Url;

use crate::{InputValueError, InputValueResult};

/// The `PossibleScalar` enum is the list of possible Scalar usable within dynaql
#[derive(Debug, Clone, Copy)]
pub enum PossibleScalar {
    String,
    Int,
    Float,
    Boolean,
    ID,
    DateTime,
    JSON,
    Url,
    Email,
}

const SPECIFIED_BY_DIRECTIVE: &str = r#"
directive @specifiedBy(url: String!) on SCALAR
"#;

const DATETIME_DIRECTIVE: &str = r#"
"""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, is compliant with the date-time format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.

This scalar is a description of an exact instant on the timeline such as the instant that a user account was created.
"""
scalar DateTime @specifiedBy(url: "https://datatracker.ietf.org/doc/html/rfc3339")
"#;

const URL_DIRECTIVE: &str = r#"
"""
A URL as defined byÂ RFC1738. For example, `https://grafbase.com/foo/` or `mailto:example@grafbase.com`.
"""
scalar URL @specifiedBy(url: "http://url.spec.whatwg.org/")
"#;

const JSON_DIRECTIVE: &str = r#"
"""
Any JSON value.
"""
scalar JSON
"#;

const EMAIL_DIRECTIVE: &str = r#"
"""
A scalar to validate the email as it is defined in the HTML specification.
"""
scalar Email @specifiedBy(url: "https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address")
"#;

impl PossibleScalar {
    /// Function to **check** if the inputed value is able to be cast into the expected type.
    /// TODO: In the future, we should also do more than just a check at the request parsing, we
    /// should also allow casting to an expected scalar to be used inside the resolving chain.
    pub(crate) fn check_valid(&self, value: &ConstValue) -> InputValueResult<bool> {
        match (self, value) {
            (Self::String | Self::ID, ConstValue::String(_)) => Ok(true),
            (Self::Boolean, ConstValue::Boolean(_)) => Ok(true),
            (Self::Int, ConstValue::Number(num)) => Ok(!num.is_f64()),
            (Self::Float, ConstValue::Number(_)) => Ok(true),
            (Self::DateTime, ConstValue::String(date)) => {
                date.parse::<DateTime<Utc>>()
                    .map_err(|err| InputValueError::ty_custom("DateTime", err))?;
                Ok(true)
            }
            (Self::Url, ConstValue::String(value)) => Url::from_str(value)
                .map_err(|err| InputValueError::ty_custom("URL", err))
                .map(|_| true),
            (Self::Email, ConstValue::String(value)) => parse_email(value)
                .map_err(|err| InputValueError::ty_custom("Email", err))
                .map(|_| true),
            (Self::JSON, ConstValue::Object(_)) => Ok(true),
            _ => Ok(false),
        }
    }

    /// Generate directives associated
    pub const fn directives() -> &'static str {
        concatcp!(
            SPECIFIED_BY_DIRECTIVE,
            '\n',
            DATETIME_DIRECTIVE,
            '\n',
            JSON_DIRECTIVE,
            '\n',
            URL_DIRECTIVE,
            '\n',
            EMAIL_DIRECTIVE
        )
    }
}

#[derive(Debug, Clone, thiserror::Error)]
pub enum PossibleScalarErrors {
    #[error("\"{expected_ty}\" is not a proper scalar")]
    NotAScalar { expected_ty: String },
}

impl TryFrom<&str> for PossibleScalar {
    type Error = PossibleScalarErrors;
    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "String" => Ok(PossibleScalar::String),
            "Int" => Ok(PossibleScalar::Int),
            "Float" => Ok(PossibleScalar::Float),
            "Boolean" => Ok(PossibleScalar::Boolean),
            "ID" => Ok(PossibleScalar::ID),
            "DateTime" => Ok(PossibleScalar::DateTime),
            "JSON" => Ok(PossibleScalar::JSON),
            "URL" => Ok(PossibleScalar::Url),
            "Email" => Ok(PossibleScalar::Email),
            _ => Err(PossibleScalarErrors::NotAScalar {
                expected_ty: value.to_string(),
            }),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::PossibleScalar;
    use dynaql_value::ConstValue;

    #[test]
    fn check_json_valid() {
        let value = serde_json::json!({
            "code": 200,
            "success": true,
            "payload": {
                "features": [
                    "serde",
                    "json"
                ]
            }
        });

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PossibleScalar::JSON.check_valid(&const_value);
        assert!(scalar.is_ok());
    }

    #[test]
    fn check_url_valid() {
        let value = serde_json::Value::String("https://grafbase.com".to_string());

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PossibleScalar::Url.check_valid(&const_value);
        assert!(scalar.is_ok());
    }

    #[test]
    fn check_url_mailto_valid() {
        let value = serde_json::Value::String("mailto:anthony@grafbase.com".to_string());

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PossibleScalar::Url.check_valid(&const_value);
        assert!(scalar.is_ok());
    }

    #[test]
    fn check_mail_valid() {
        let value = serde_json::Value::String("anthony@grafbase.com".to_string());

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PossibleScalar::Email.check_valid(&const_value);
        assert!(scalar.is_ok());
    }

    #[test]
    fn check_not_mail_valid() {
        let value = serde_json::Value::String("anthony @grafbase.com".to_string());

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PossibleScalar::Email.check_valid(&const_value);
        assert!(scalar.is_err());
    }
}
