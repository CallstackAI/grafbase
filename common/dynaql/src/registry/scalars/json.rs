use super::{DynamicParse, SDLDefinitionScalar};
use crate::{Error, InputValueError, InputValueResult};
use dynaql_value::ConstValue;

pub struct JSONScalar;

impl<'a> SDLDefinitionScalar<'a> for JSONScalar {
    fn name() -> Option<&'a str> {
        Some("JSON")
    }

    fn description() -> Option<&'a str> {
        Some(
            r#"
            A JSON Value
            "#,
        )
    }
}

impl DynamicParse for JSONScalar {
    fn is_valid(value: &ConstValue) -> bool {
        matches!(value, ConstValue::Object(_))
    }

    fn to_value(value: serde_json::Value) -> Result<ConstValue, Error> {
        match value {
            val @ serde_json::Value::Object(_) => {
                ConstValue::from_json(val).map_err(Error::new_with_source)
            }
            _ => Err(Error::new(
                "Data violation: Cannot coerce the initial value to a JSON",
            )),
        }
    }

    fn parse(value: ConstValue) -> InputValueResult<serde_json::Value> {
        match value {
            val @ ConstValue::Object(_) => {
                ConstValue::into_json(val).map_err(|err| InputValueError::ty_custom("JSON", err))
            }
            _ => Err(InputValueError::ty_custom(
                "JSON",
                "Cannot parse into a JSON",
            )),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::registry::scalars::{DynamicParse, JSONScalar};
    use dynaql_value::ConstValue;

    #[test]
    fn check_json_valid() {
        let value = serde_json::json!({
            "code": 200,
            "success": true,
            "payload": {
                "features": [
                    "serde",
                    "json"
                ]
            }
        });

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = JSONScalar::parse(const_value);
        assert!(scalar.is_ok());
    }
}
