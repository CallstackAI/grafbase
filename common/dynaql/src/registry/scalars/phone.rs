use super::{DynamicParse, SDLDefinitionScalar};
use crate::{Error, InputValueError, InputValueResult};
use dynaql_value::ConstValue;
use phonenumber as _;

// TODO: Input coercion to accept either ms or a date
pub struct PhoneScalar;

impl<'a> SDLDefinitionScalar<'a> for PhoneScalar {
    fn name() -> Option<&'a str> {
        Some("Phone")
    }

    fn description() -> Option<&'a str> {
        Some("A phone number. This value is stored as a string. Phone numbers can contain either spaces or hyphens to separate digit groups. Phone numbers without a country code are assumed to be US/North American numbers adhering to theÂ North American Numbering Plan (NANP).")
    }
}

impl DynamicParse for PhoneScalar {
    fn is_valid(value: &ConstValue) -> bool {
        match value {
            ConstValue::String(phone) => phonenumber::parse(None, phone)
                .map(|phone| phone.is_valid())
                .unwrap_or_default(),
            _ => false,
        }
    }

    fn to_value(value: serde_json::Value) -> Result<ConstValue, Error> {
        match value {
            serde_json::Value::String(phone) => Ok(ConstValue::String(phone)),
            _ => Err(Error::new(
                "Data violation: Cannot coerce the initial value into an Phone",
            )),
        }
    }

    fn parse(value: ConstValue) -> InputValueResult<serde_json::Value> {
        match value {
            ConstValue::String(phone) => {
                let e164_phone = phonenumber::parse(None, phone)
                    .map_err(|err| InputValueError::ty_custom("Phone", err))?;

                if e164_phone.is_valid() {
                    Ok(serde_json::Value::String(
                        e164_phone
                            .format()
                            .mode(phonenumber::Mode::International)
                            .to_string(),
                    ))
                } else {
                    Err(InputValueError::ty_custom("Phone", "Invalid phone number"))
                }
            }
            _ => Err(InputValueError::ty_custom(
                "Phone",
                "Cannot parse into a Phone",
            )),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::super::SDLDefinitionScalar;
    use crate::registry::scalars::{DynamicParse, PhoneScalar};
    use dynaql_value::ConstValue;
    use insta::assert_snapshot;

    #[test]
    fn check_test_phonenumber() {
        let value = serde_json::Value::String("+33612121212".to_string());

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PhoneScalar::parse(const_value);
        assert!(scalar.is_ok());
    }

    #[test]
    fn check_test_fail_phonenumber() {
        let value = serde_json::Value::String("+3361212121".to_string());

        let const_value = ConstValue::from_json(value).unwrap();

        let scalar = PhoneScalar::parse(const_value);
        assert!(scalar.is_err());
        insta::assert_debug_snapshot!(scalar);
    }

    #[test]
    fn ensure_directives_sdl() {
        assert_snapshot!(PhoneScalar::sdl());
    }
}
