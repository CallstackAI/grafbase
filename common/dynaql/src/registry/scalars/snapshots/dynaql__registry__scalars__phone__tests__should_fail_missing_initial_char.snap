---
source: common/dynaql/src/registry/scalars/phone.rs
assertion_line: 102
expression: scalar
---
Err(
    InputValueError {
        message: "Failed to parse \"PhoneNumber\": Invalid phone number",
        phantom: PhantomData<serde_json::value::Value>,
    },
)
